File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        
        Arrays.sort(nums);
        int n = nums.length;
        List<List<Integer>>ans = new ArrayList<>();
        for(int i = 0 ; i < n ; i++)
        {
            if(i > 0 && nums[i] == nums[i-1])
                continue;
            int num = nums[i];
            int target = -nums[i];
            int start = i+1;
            int end = n-1;
            while(start < end)
            {
                int sum = nums[start] + nums[end];
                if(sum == target)
                {
                    List<Integer>list1 = new ArrayList<>(List.of(nums[i], nums[start], nums[end]));
                    ans.add(list1);
                    while(start < end && nums[start] == nums[start+1])
                    {
                        start++;
                    }
                    while(start < end && nums[end] == nums[end-1])
                    {
                        end--;
                    }
                    start++;
                    end--;
                }
                else if(sum < target)
                    start++;
                else
                    end--;
            }
        }
        return ans;
    }
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:

        ans = []
        nums.sort()
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            start = i + 1
            end = len(nums) - 1
            while start < end :
                sum_total = nums[i] + nums[start] + nums[end]
                
                if sum_total < 0 :
                    start+=1
                elif sum_total > 0 :
                    end-=1
                elif sum_total == 0:
                    list1 = [nums[i], nums[start], nums[end]]
                    ans.append(list1)
                    start+=1
                    while start < end and nums[start] == nums[start-1]:
                        start+=1
        return ans

